nnvm/python/nnvm/testing/check_computation.py

```diff
@@ -0,0 +1,451 @@
+
+        debug_stage = None
+
+        try:
+            debug_stage = "compiling"
+            nnvm_res = None
+            main_function = graph_to_function(main_graph, target, ctx)
+            # nnvm_res contains the output and gradients (if they are needed)
+            nnvm_res = main_function(**np_inputs)
+
+            if np_forward is not None:
+                debug_stage = "checking forward computation"
+                logging.debug(debug_stage)
+
+                numpy_res = np_forward(**np_inputs_without_head_grads)
+                np.testing.assert_allclose(nnvm_res[0], numpy_res, atol=atol, rtol=rtol)
+
+            if np_backward is not None:
+                debug_stage = "checking gradients"
+                logging.debug(debug_stage)
+
+                numpy_grads = np_backward(**np_inputs)
+                for i, np_grad in enumerate(numpy_grads):
+                    np.testing.assert_allclose(nnvm_res[i + 1], np_grad, atol=atol, rtol=rtol)
+
+            if numerical_grads:
+                debug_stage = "checking gradients numerically"
+                logging.debug(debug_stage)
+
+                forward_function = graph_to_function(forward_graph, target, ctx)
+
+                # Since the result may be non-scalar, we have to put another operation on the top,
+                # so we just multiple by the randomly generated head_grads.
+                # This way we can reuse the gradient values which has been already computed.
+                def function(**kwargs):
+                    res = forward_function(**kwargs)[0]
+                    return np.dot(np_inputs['head_grads'].ravel(), res.ravel())
+
+                function_value = np.dot(np_inputs['head_grads'].ravel(), nnvm_res[0].ravel())
+                grad_var_names = [x.attr('name') for x in grad_input_vars]
+                grad_values = {x: v for x, v in zip(grad_var_names, nnvm_res[1:])}
+
+                check_numerical_grads(
+                    function,
+                    grad_input_vars=grad_var_names,
+                    input_values=np_inputs_without_head_grads,
+                    function_value=function_value,
+                    grad_values=grad_values,
+                    delta=delta,
+                    max_error=ng_max_error,
+                    max_discarded_frac=ng_max_discarded_frac,
+                    atol=ng_atol, rtol=ng_rtol)
+
+        except:
+            print("\ncheck_function failed while {}, here is the main graph".format(debug_stage))
+            print(main_graph.ir(join_node_attrs=['shape', 'dtype']))
+            if nnvm_res is not None:
+                print("Generated inputs:")
+                print(np_inputs)
+                print()
+            raise
+
+
+def check_numerical_grads(function, grad_input_vars, input_values, grad_values, function_value=None,
+                          delta=1e-3, max_error=1e+3, max_discarded_frac=0.1,
```

Comment:
The default value of delta is different from the check_function's one.  Should be 1e-2?

Reply:
Oh, it should be 1e-3 everywhere, thank you!